egs/wsj/s5/steps/cleanup/segment_long_utterances.sh

```diff
@@ -0,0 +1,389 @@
+#!/bin/bash
+
+# Copyright 2014  Guoguo Chen
+#           2016  Vimal Manohar
+# Apache 2.0
+
+set -e 
+set -o pipefail
+
+. path.sh
+
+set -u
+
+# Uniform segmentation options
+max_segment_duration=30
+overlap_duration=5
+
+# Decode options
+graph_opts=
+nj=4
+lmwt=10
+
+# TF-IDF similarity search options
+max_words=1000
+num_neighbors_to_search=1
+neighbor_tfidf_threshold=0.5
+
+# First-pass segmentation opts
+min_segment_length=0.5
+min_new_segment_length=1.0
+max_tainted_length=0.05
+max_edge_silence_length=0.5
+max_edge_non_scored_length=0.5
+max_internal_silence_length=2.0
+max_internal_non_scored_length=2.0
+unk_padding=0.05
+max_junk_proportion=0.1
+min_split_point_duration=0.1
+max_deleted_words_kept_when_merging=1
+silence_factor=1
+incorrect_words_factor=1
+tainted_words_factor=1
+max_wer=50
+max_segment_length_for_merging=60
+max_bad_proportion=0.75
+max_intersegment_incorrect_words_length=1
+max_segment_length_for_splitting=10
+hard_max_segment_length=15
+min_silence_length_to_split_at=0.3
+min_non_scored_length_to_split_at=0.3
+
+stage=-1
+
+cmd=run.pl
+
+. utils/parse_options.sh
+
+if [ $# -ne 5 ]; then
+    cat <<EOF
+Usage: $0 [options] <model-dir> <lang> <data-in> <segmented-data-out> <work-dir>
+ e.g.: $0 exp/tri3 data/train_unsegmented data/train_segmented exp/tri3_segmentation
+This script performs segmentation of the data in <data-in>, writing the segmented
+data (with a segments file) to <segmented-data-out>.  The purpose of this script is
+to divide up the input data (which may consist of long recordings such as television
+shows or audiobooks) into segments which are of manageable length for further
+processing, along with the portion of the transcript that seems to match each segment.
+The output data is not necessarily particularly clean; you are advised to 
+run steps/cleanup/clean_and_segment_data.sh on the output in order to further
+clean it and eliminate data where the transcript doesn't seem to match.
+EOF
+    exit 1
+fi
+
+srcdir=$1
+lang=$2
+data=$3
+out_data=$4
+dir=$5
+
+for f in $data/feats.scp $data/cmvn.scp $srcdir/tree \
+  $srcdir/final.mdl $srcdir/cmvn_opts; do
+  if [ ! -f $f ]; then
+    echo "$0: Could not find file $f"
+    exit 1
+  fi
+done
+
+data_id=`basename $data`
+mkdir -p $dir
+
+data_uniform_seg=$dir/${data_id}_uniform_seg
+
+# First we split the data into segments of around 30s long, on which 
+# it would be possible to do a decoding.
+if [ $stage -le 1 ]; then
+  echo "$0: Splitting data directory $data into uniform segments..."
+
+  utils/data/get_utt2dur.sh $data
+
+  # TODO: Write this script
+  # utils/data/get_uniform_subsegments.py \
+  #   --max-segment-duration=$max_segment_duration \
+  #   --overlap-duration=$overlap_duration \
+  #   $data/segments > $dir/uniform_sub_segments
+
+  utils/data/get_segments_for_data.sh $data > $data/segments
+
+  cat $data/utt2spk | awk 'BEGIN{i=1} {print $1" "i; i++}' > $data/utt2label
+
+  segmentation-init-from-segments --shift-to-zero=true \
```

Comment:
I think it's better to use python for now, at least until the bigger picture of diarization is decided. Each new C++ program adds compilation overhead.